lib/rspec-puppet-facts.rb

```diff
@@ -148,26 +148,16 @@ def on_supported_os_implementation(opts = {})
 
     os_facts_hash = {}
     received_facts.map do |facts|
-      # Fix facter bug
-      # Todo: refactor the whole block to rely on structured facts and use legacy facts as fallback
-      if facts[:operatingsystem] == 'Ubuntu'
-        operatingsystemmajrelease = facts[:operatingsystemrelease].split('.')[0..1].join('.')
-      elsif facts[:operatingsystem] == 'OpenBSD'
-        operatingsystemmajrelease = facts[:operatingsystemrelease]
-      elsif facts[:operatingsystem] == 'windows' && facts[:operatingsystemrelease].start_with?('10.0.')
-        operatingsystemmajrelease = '2016'
-      elsif facts.dig(:os, 'release', 'major')
-        operatingsystemmajrelease = facts[:os]['release']['major']
-      elsif facts.dig(:os, 'distro', 'release', 'major')
-        operatingsystemmajrelease = facts[:os]['distro']['release']['major']
+      if RSpec.configuration.facterdb_string_keys
+        operatingsystemmajrelease = facts['os']['release']['major']
+        operatingsystem = facts['os']['name'].downcase
+        hardwaremodel = facts['os']['hardware']
       else
-        if facts[:operatingsystemmajrelease].nil?
-          operatingsystemmajrelease = facts[:operatingsystemrelease].split('.')[0]
-        else
-          operatingsystemmajrelease = facts[:operatingsystemmajrelease]
-        end
+        operatingsystemmajrelease = facts[:os]['release']['major']
+        operatingsystem = facts[:os]['name'].downcase
+        hardwaremodel = facts[:os]['hardware']
       end
-      os = "#{facts[:operatingsystem].downcase}-#{operatingsystemmajrelease}-#{facts[:hardwaremodel]}"
+      os = "#{operatingsystem}-#{operatingsystemmajrelease}-#{hardwaremodel}"
       next unless os.start_with? RspecPuppetFacts.spec_facts_os_filter if RspecPuppetFacts.spec_facts_os_filter
       facts.merge! RspecPuppetFacts.common_facts
       os_facts_hash[os] = RspecPuppetFacts.with_custom_facts(os, facts)
```

Comment:
Just below is a call to `stringify_keys`, which I think is redundant if you ask FacterDB to give you strings.

Reply:
yes, done